controller/controller.go

```diff
@@ -16,26 +21,144 @@ import (
+	// A nested map of endpoint weights for a language pair.
+	// The first key represents the source language, the second key represents the target language, the 
+	// third key represents the endpoint URL and the value the corresponding weight for that endpoint.
+	LanguagePairWeights map[string]map[string]map[string]float64
+}
+
+// NewLnxEndpointConfiguration returns a new endpoint configuration based on a list of endpoints, weights and list of supported languages
+func NewLnxEndpointConfiguration(endpoints []string, weights []float64, languageLists []language.GoogleLanguageList) (*LnxEndpointConfiguration, error) {
+	if len(endpoints) != len(weights) || len(weights) != len(languageLists) {
+		return nil, fmt.Errorf("Number of endpoints must match number of weights and number of language lists")
+	}
+
+	conf := LnxEndpointConfiguration{
+		Endpoints:           endpoints,
+		DefaultWeights:      weights,
+		LanguagePairList:    language.GoogleLanguageList{Sl: make(map[string]string), Tl: make(map[string]string)},
+		LanguagePairWeights: make(map[string]map[string]map[string]float64),
+	}
+
+	for i, endpoint := range endpoints {
+		// get the list of supported languages for the current endpoint
+		list := languageLists[i]
+
+		// iterate through the source languages the current endpoint supports
+		for sl, sldesc := range list.Sl {
+			// add the source language description to the merged language pair list
+			conf.LanguagePairList.Sl[sl] = sldesc
+
+			// check if the source language weight map already exists in the language pair weights
+			if _, ok := conf.LanguagePairWeights[sl]; !ok {
+				// if not, create a new weight map for the source language
+				conf.LanguagePairWeights[sl] = make(map[string]map[string]float64)
+			}
+
+			for tl, tldesc := range list.Tl {
+				// add the target language description to the merged language pair list
+				conf.LanguagePairList.Tl[tl] = tldesc
+
+				// check if the weight map for the source / target language pair already exists
+				if _, ok := conf.LanguagePairWeights[sl][tl]; !ok {
+					// if not, create a new weight map for it
+					conf.LanguagePairWeights[sl][tl] = make(map[string]float64)
+				}
+				 // set the default weight for the current endpoint for the source-target language pair
+				conf.LanguagePairWeights[sl][tl][endpoint] = conf.DefaultWeights[i]
+			}
+		}
+	}
+	return &conf, nil
+}
+
+// GetEndpoint returns the endpoint which should be used based on the weights and languages supported.
+func (c *LnxEndpointConfiguration) GetEndpoint(from, to string) string {
+	// initialize total weight and incrementals.
+	total := 0.0
+	incrementals := []float64{}
+
+	// retrieve the nested map of language pair weights.
+	weights := c.LanguagePairWeights[from][to]
+
+	// iterate through the Endpoints array, accumulating the total weight and storing the intermediate sums in incrementals.
+	for _, endpoint := range c.Endpoints {
+		total += weights[endpoint]
```

Comment:
When we going to use a lot of endpoints, better to calc `total ` and `incrementals` once.
For the current config it doesn't matter

Reply:
we'll only use 2 endpoints in practice so this shouldn't matter much